runtime/memprof.c

```diff
@@ -880,7 +880,8 @@ void caml_memprof_track_young(uintnat wosize, int from_caml,
       t_idx = new_tracked(n_samples, alloc_wosz, SRC_NORMAL, 1,
                           Placeholder_offs(alloc_ofs), callstack);
       if (t_idx == Invalid_index) continue;
-      res = run_alloc_callback_exn(t_idx);
+      if (Caml_state->mask_async_callbacks == CAML_MASK_NONE)
+        res = run_alloc_callback_exn(t_idx);
```

Comment:
If I understand correctly, the rest of the loop-iteration code below is useless if the callback is not immediately run, so you should be able to just `continue` in the masked case.

Note: to me checking `== CAML_MASK_NONE` corresponds to a fragile assumption the mask lattice, that there is no intermediary mask between NONE and masks_memprof_callbacks. I would rather write the code as `< memprof-callback-mask`.

Reply:
Except for the `if (!stopped) ...` at the end, right?

`memprof.c` and `signals.c` already have tightly coupled assumptions, which is why it is fine to test `== CAML_MASK_NONE`.

Reply:
I now understood that you meant replacing `== CAML_MASK_NONE` with `< CAML_MASK_UNINTERRUPTIBLE`, not introducing a new variable for generality.

Reply:
> Except for the `if (!stopped) ...` at the end, right?

Yes, you are right, my suggestion above changes the behavior compared to your patch. But I am not sure that either one (your patch or my suggestion) is correct, and I find it difficult to understand what should be done here. For example, with your version, `res` is not updated when the mask is set, so it could retain an exceptional value that it got from a previous iteration of the loop, and `break` on line 897 due to an old/stale value. Now we would normally assume that the mask will not be set during this loop (or locally inside a callback and unset again before the next iteration), I don't know if this can happen.

If I understand correctly, what you want to do about the interaction of masking of memprof is to keep sampling as usual, just not run the callbacks. This sounds sensible but I am skeptical that the current implementation is correct (or clear enough for me to check its correctness). We should probably ask @jhjourdan or @stedolan to look at this part of the change.

Reply:
Yes, the reasoning is that this way it behaves similarly to the call to `maybe_track_block` in the case of an alloc from C, near the beginning of the function.

`res` is not an issue: it cannot be an exceptional value or it would have exited the loop (and although it is not rooted, there is never any heap allocation between the moment it is set and the moment it is read).

We are allowed to assume that calls to `caml_mask`/`caml_unmask` are well bracketed, but I do not think this is relevant here.